### PR TITLE
feat(map): Codegen for XMLSchema Target

### PR DESCRIPTION
**Pull request by Jonathan-Casey**
### Description
Add XMLSchema code generation for Map Type. Validated against [xmlvalidation.com](https://www.xmlvalidation.com/) &  [w3schools.com](https://www.w3schools.com/xml/xml_validator.asp/).

#### Sample <String, String>
```
<xs:complexType name="Map1">
   <xs:sequence>
      <xs:complexType>
         <xs:sequence>
            <xs:element name="key" type="xs:string"/>
            <xs:element name="value" type="xs:string"/>
         </xs:sequence>
      </xs:complexType>
   </xs:sequence>
</xs:complexType>
```

#### Sample <String, DateTime>
```
<xs:complexType name="Map2">
   <xs:sequence>
      <xs:complexType>
         <xs:sequence>
            <xs:element name="key" type="xs:string"/>
            <xs:element name="value" type="xs:dateTime"/>
         </xs:sequence>
      </xs:complexType>
   </xs:sequence>
</xs:complexType>
```

#### Sample <String, SSN>
```
<xs:complexType name="Map3">
   <xs:sequence>
      <xs:complexType>
         <xs:sequence>
            <xs:element name="key" type="xs:string"/>
            <xs:element name="value" type="xs:string"/>
         </xs:sequence>
      </xs:complexType>
   </xs:sequence>
</xs:complexType>
```

#### Sample <String, Concept>
```
<xs:complexType name="Map4">
   <xs:sequence>
      <xs:complexType>
         <xs:sequence>
            <xs:element name="key" type="xs:string"/>
            <xs:element name="value" type="concerto:Concept"/>
         </xs:sequence>
      </xs:complexType>
   </xs:sequence>
</xs:complexType>
```

#### Sample <SSN, String>
```
<xs:complexType name="Map5">
   <xs:sequence>
      <xs:complexType>
         <xs:sequence>
            <xs:element name="key" type="xs:string"/>
            <xs:element name="value" type="xs:string"/>
         </xs:sequence>
      </xs:complexType>
   </xs:sequence>
</xs:complexType>
```

#### Sample <SSN, Employee>
```
<xs:complexType name="Map6">
   <xs:sequence>
      <xs:complexType>
         <xs:sequence>
            <xs:element name="key" type="xs:string"/>
            <xs:element name="value" type="org.acme.hr:Employee"/>
         </xs:sequence>
      </xs:complexType>
   </xs:sequence>
</xs:complexType>
```

### CODE REVIEW

File: lib/codegen/fromcto/xmlschema/xmlschemavisitor.js [hunk 1]
```diff
@@ -153,6 +153,33 @@ class XmlSchemaVisitor {
         return null;
     }
 
+    /**
+     * Visitor design pattern
+     * @param {MapDeclaration} mapDeclaration - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
```
Comment (ekarademir) [inline lib/codegen/fromcto/xmlschema/xmlschemavisitor.js:160]: This function only returns `null`?

**Review by ekarademir — APPROVED** [review]

**Review by mttrbrts — COMMENTED** [review]
I had trouble validating the output schema. I'm also unsure about what the target XML document would look like for instances.

For example, the following fails to validate in VS Code.
```xml
<?xml version="1.0"?>
<xs:schema xmlns:org.acme.hr="org.acme.hr" targetNamespace="org.acme.hr" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema" 
           xmlns:concerto="concerto"
  >
  <xs:element name="root">
    <xs:complexType name="Map1">
      <xs:sequence>
        <xs:complexType>
          <xs:sequence>
            <xs:element name="key" type="xs:string"/>
            <xs:element name="value" type="xs:string"/>
          </xs:sequence>
        </xs:complexType>
      </xs:sequence>
    </xs:complexType>
  </xs:element>
</xs:schema>
```

Problems:
- s4s-att-not-allowed: Attribute 'name' cannot appear in element 'complexType'.
- s4s-elt-must-match.1: The content of 'sequence' must match (annotation?, (element | group | choice | sequence | any)*). A problem was found starting at: complexType.

If I change the definition to the following, it does validate, but produces unexpected instances.

```
<?xml version="1.0"?>
<xs:schema xmlns:org.acme.hr="org.acme.hr" targetNamespace="org.acme.hr" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema" 
           xmlns:concerto="concerto"
  >
  <xs:element name="root">
    <xs:complexType>
      <xs:sequence>
        <xs:element name="key" type="xs:string"/>
        <xs:element name="value" type="xs:string"/>
      </xs:sequence>
    </xs:complexType>
  </xs:element>
</xs:schema>
```

```
<?xml version="1.0" encoding="utf-8"?>
<!-- Created with Liquid Technologies Online Tools 1.0 (https://www.liquid-technologies.com) -->
<root xmlns="org.acme.hr" xmlns:concerto="concerto" xsi:schemaLocation="org.acme.hr schema.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <key>string</key>
  <value>1981-11-16T18:39:36.34</value>
</root>
```

Is the [design here ](https://stackoverflow.com/questions/35342044/xml-schema-for-representing-key-value-pairs) better suited to our use case?

**Comment by Jonathan-Casey** [comment]
> I had trouble validating the output schema. I'm also unsure about what the target XML document would look like for instances.
> 
> For example, the following fails to validate in VS Code.
> 
> ```
> <?xml version="1.0"?>
> <xs:schema xmlns:org.acme.hr="org.acme.hr" targetNamespace="org.acme.hr" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema" 
>            xmlns:concerto="concerto"
>   >
>   <xs:element name="root">
>     <xs:complexType name="Map1">
>       <xs:sequence>
>         <xs:complexType>
>           <xs:sequence>
>             <xs:element name="key" type="xs:string"/>
>             <xs:element name="value" type="xs:string"/>
>           </xs:sequence>
>         </xs:complexType>
>       </xs:sequence>
>     </xs:complexType>
>   </xs:element>
> </xs:schema>
> ```
> 
> Problems:
> 
> * s4s-att-not-allowed: Attribute 'name' cannot appear in element 'complexType'.
> * s4s-elt-must-match.1: The content of 'sequence' must match (annotation?, (element | group | choice | sequence | any)*). A problem was found starting at: complexType.
> 
> If I change the definition to the following, it does validate, but produces unexpected instances.
> 
> ```
> <?xml version="1.0"?>
> <xs:schema xmlns:org.acme.hr="org.acme.hr" targetNamespace="org.acme.hr" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema" 
>            xmlns:concerto="concerto"
>   >
>   <xs:element name="root">
>     <xs:complexType>
>       <xs:sequence>
>         <xs:element name="key" type="xs:string"/>
>         <xs:element name="value" type="xs:string"/>
>       </xs:sequence>
>     </xs:complexType>
>   </xs:element>
> </xs:schema>
> ```
> 
> ```
> <?xml version="1.0" encoding="utf-8"?>
> 
> <root xmlns="org.acme.hr" xmlns:concerto="concerto" xsi:schemaLocation="org.acme.hr schema.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
>   <key>string</key>
>   <value>1981-11-16T18:39:36.34</value>
> </root>
> ```
> 
> Is the [design here ](https://stackoverflow.com/questions/35342044/xml-schema-for-representing-key-value-pairs) better suited to our use case?

@mttrbrts thanks for catching this. I've switched validators and was able to reproduce the same issues which are now patched. 

RedHats VSCode plugin used for inline validation and [liquid-technologies ](https://www.liquid-technologies.com/online-xsd-validator)online instance validator both suggest the current generation of XSD is valid. Here is a sample:

**XSD:** 
```
<?xml version="1.0"?>
<xs:schema xmlns:org.acme.hr="org.acme.hr" targetNamespace="org.acme.hr"
    elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xmlns:concerto="concerto"
>
       <xs:complexType name="CompanyProperties">
        <xs:sequence>
           <xs:element name="key" type="xs:string"/>
           <xs:element name="value" type="xs:string"/>
        </xs:sequence>
     </xs:complexType>
     <xs:element name="CompanyProperties" type="org.acme.hr:CompanyProperties"/>
</xs:schema>

```

**Instance:**

```
<?xml version="1.0" encoding="UTF-8"?>
<org.acme.hr:CompanyProperties xmlns:org.acme.hr="org.acme.hr">
  <org.acme.hr:key>Some Thing</org.acme.hr:key>
  <org.acme.hr:value>Some Other Thing</org.acme.hr:value>
</org.acme.hr:CompanyProperties>
```

**Comment by Jonathan-Casey** [comment]
A sample XSD which supports multiple entries in a Map. Each Map will have its own 'entry' type defining the entries within the instance.

```
<xs:element name="diary">
  <xs:complexType>
    <xs:sequence>
      <xs:element name="entry" type="diary_entryType" minOccurs="0" maxOccurs="unbounded"/>
    </xs:sequence>
  </xs:complexType>
</xs:element>
```

```
<xs:complexType name="diary_entryType">
  <xs:sequence>
    <xs:element name="key" type="xs:xs:dateTime"/>
    <xs:element name="value" type="xs:string"/>
  </xs:sequence>
</xs:complexType>
```

**Review by dselman — APPROVED** [review]

**Review by mttrbrts — APPROVED** [review]

**Pull request merged**